qiling/debugger/gdb/gdb.py

```diff
@@ -3,811 +3,973 @@
+        """Iterate through incoming packets in an active connection until
+        it is terminated.
+        """
+
+        pattern = re.compile(br'^\$(?P<data>[^#]*)#(?P<checksum>[0-9a-fA-F]{2})')
+        buffer = bytearray()
+
+        while True:
+            try:
+                incoming = self.client.recv(self.BUFSIZE)
+            except ConnectionError:
+                break
+
+            # remote connection closed
+            if not incoming:
                 break
 
-            if cmd not in commands:
-                self.send('')
-                self.ql.log.info("gdb> Command not supported: %s\n" %(cmd))
+            buffer += incoming
+
+            # discard incoming acks
+            if buffer[0:1] == REPLY_ACK:
+                del buffer[0]
+
+            packet = pattern.match(buffer)
+
+            # if there is no match, the rest of the packet might be missing
+            if not packet:
                 continue
-            self.ql.log.debug("gdb> received: %s%s" % (cmd, subcmd))
-            commands[cmd](subcmd)
-
-        self.close()
-
-
-    def receive(self):
-        '''Receive a packet from a GDB client'''
-        csum = 0
-        state = 'Finding SOP'
-        packet = ''
-        try:
-            while True:
-                c = self.netin.read(1)
-                if c == '\x03':
-                    return 'Error: CTRL+C'
-
-                if len(c) != 1:
-                    return 'Error: EOF'
-
-                if state == 'Finding SOP':
-                    if c == '$':
-                        state = 'Finding EOP'
-                elif state == 'Finding EOP':
-                    if c == '#':
-                        if csum != int(self.netin.read(2), 16):
-                            raise Exception('invalid checksum')
-                        self.last_pkt = packet
-                        return 'Good'
-                    else:
-                        packet += c
-                        csum = (csum + ord(c)) & 0xff
-                else:
-                    raise Exception('should not be here')
-        except:
-            self.close()
-            raise
-
-    def checksum(self, data):
-        checksum = 0
-        for c in data:
-            if type(c) == str:
-                checksum += (ord(c))
-            else:
-                checksum += c
-        return checksum & 0xff
 
-    def send(self, msg):
-        """Send a packet to the GDB client"""
-        if type(msg) == str:
-            self.send_raw('$%s#%.2x' % (msg, self.checksum(msg)))
+            data = packet['data']
+            read_csum = int(packet['checksum'], 16)
+            calc_csum = GdbSerialConn.checksum(data)
+
+            if read_csum != calc_csum:
+                raise IOError(f'checksum error: expected {calc_csum:02x} but got {read_csum:02x}')
+
+            # follow gdbserver debug output format
+            self.log.debug(f'getpkt ("{GdbSerialConn.__printable_prefix(data).decode(ENCODING)}");')
```

Comment:
This is too verbose, could we have a more verbose logging level? Sometimes users turn on debug level just to see some extra information of the emulation not the packets.

Reply:
I agree, but at the same time I wanted to follow gdbserver debugging messages - as the comment there says.
The previous version of the server also printed out a message upon each incoming packet (though it didn't emit the payload..). Adding an excessive verbosity level is a great idea.. but what do you suggest to do in the meantime?

Reply:
I suggest adding a `QL_VERBOSE.TRACE` with a value of 5 (`logging.DEBUG` is 10) or so. This might not the goal of this PR, but I think it's a good time to do it. ;p

Reply:
What is  your suggestion until we do?
Disable the getpkt / putpkt messages, or leave it as-is?

Reply:
The debug messages are indeed helpful so I would like to leave it as-is. I would draft another PR for `QL_VERBOSE.TRACE`.

Reply:
I have something a bit different in mind for further verbosity levels.
Let's discuss that before you start working on this.